hibernate-core/src/main/java/org/hibernate/persister/entity/UnionSubclassEntityPersister.java

```diff
@@ -222,6 +222,11 @@ public String getTableName() {
 		return subquery;
 	}
 
+	@Override
+	public String[] getTableNames() {
+		return new String[] {tableName};
```

Comment:
I think you do not have to be afraid about `UnionSubclassEntityPersister`. It totally ignores `referencedTables` collection. Now, when the implementation of getTableNames() is in `AbstractEntityPersister`, `UnionSubclassEntityPersister` has no change in comparison with master branch.

Only `JoinedSubclassEntityPersister` use this information to optimize table joining.

Reply:
> I am not only worried about this issue with respect to subclass tables, but also for example in the context of secondary tables. I assume that if you only select a property from a secondary table, this will be the only table joined? That would be slightly odd behaviour. Can you for example add a test of entities A and B (not sharing the same root class) that reference a secondary table C, and then show the produced SQL and output for SELECT a.somePropOnC FROM A a? I assume this would now return somePropOnC values for B too, which is obviously wrong.

I added test for that, see `OmitAncestorJoinWithSecondaryTableTest`.